### PR TITLE
Improved method for decoding file name

### LINKED ISSUE #2875: Improve and beautify getDecodedFileName method
**Describe the bug**
Currently method is written as 
```
getDecodedFileName(url: String?, src: String?): String =
    url?.substringAfterLast("/", "")
      ?.takeIf { it.contains(".") }
      ?: src?.substringAfterLast("/", "")
        ?.substringAfterLast("%3A") ?: ""
```

**Expected behavior**
We can beautify and improve this piece of method code.

### PR DESCRIPTION
**Pull request by MohitMaliFtechiz**
Fixes #2875 

We are using ```URLUtils.guessFileName()``` method of ```android.webkit``` . 

* This method returns a file name guess from the url using ```URLUtils.guessFileName()``` method of ```android.webkit```. which is using ```Uri.decode``` method to extract the filename from url. After that it splits filename between base and extension (e.g for DemoFile.png,  DemoFile is base and png is extension). if there is no extension in url then it will automatically add the .bin extension to filename.
   
* If it's failed to guess the file name then it will return default filename ```downloadfile.bin```. If it returns this default value or containing the .bin in file name, then we are returning null from this function which is handled in ```downloadFileFromUrl method```.
   
* We are placing a condition here for if the file name does not have a .bin extension, then it returns the original file name.

* I have improved the test cases for this function.

### CODE REVIEW

**Review by gouri-panda — COMMENTED** [review]
@MohitMaliFtechiz The underline algorithm is still the same. As I recall, we discussed creating a new file name algorithm.
cc @kelson42

**Comment by MohitMaliFtechiz** [comment]
> @MohitMaliFtechiz The underline algorithm is still the same. As I recall, we discussed creating a new file name algorithm. cc @kelson42

hi @gouri-panda ,
 i have changed the underline algoritham for decoding file name. i have used ```URLUtil.guessFileName()``` method of ```android.webkit```.

**Comment by MohitMaliFtechiz** [comment]
hi @gouri-panda ,
now this PR is ready for review.

File: core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt [hunk 1]
```diff
@@ -315,11 +316,14 @@ object FileUtils {
    * if it contains '.'. If the url is null then it'll get the file name from the last '/'.
    * If the url and src doesn't exist it returns the empty string.
    */
-  fun getDecodedFileName(url: String?, src: String?): String =
```
Comment (gouri-panda) [inline core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt:318]: The documentation is outdated now :)

**Review by gouri-panda — COMMENTED** [review]
@MohitMaliDeveloper Just a small change then we are good to merge.

File: core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt (hunk 1 above)
Comment (MohitMaliFtechiz) [reply core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt]: hi @gouri-panda , done

**Review by gouri-panda — DISMISSED** [review]

**Comment by kelson42** [comment]
@gouri-panda A full month without review looks like :(

**Comment by MohitMaliFtechiz** [comment]
hi @gouri-panda , 
can you please re-review this PR.

**Review by gouri-panda — APPROVED** [review]

**Comment by MohitMaliFtechiz** [comment]
hi @gouri-panda ,
now you can re-review this PR.

File: core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt [hunk 2]
```diff
@@ -310,16 +311,14 @@ object FileUtils {
     return null
   }
 
-  /**
```
Comment (kelson42) [inline core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt:313]: Why have you removed the comment about what does the function? It should not IMO.

File: core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt [hunk 3]
```diff
@@ -310,16 +311,14 @@ object FileUtils {
...
-   * Returns the file name from the url or src. In url it gets the file name from the last '/' and
-   * if it contains '.'. If the url is null then it'll get the file name from the last '/'.
-   * If the url and src doesn't exist it returns the empty string.
-   */
-  fun getDecodedFileName(url: String?, src: String?): String =
-    url?.substringAfterLast("/", "")
-      ?.takeIf { it.contains(".") }
-      ?: src?.substringAfterLast("/", "")
-        ?.substringAfterLast("%3A") ?: ""
+  fun getDecodedFileName(url: String?, src: String?): String {
```
Comment (kelson42) [inline core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt:314]: I see not reason to take two arguments, considering that they are (and should be treated the same manner).

File: core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt [hunk 4]
```diff
@@ -310,16 +311,14 @@ object FileUtils {
...
+    var fileName = ""
+    val decodedFileName = URLUtil.guessFileName(url ?: src, null, null)
+    if (decodedFileName != "downloadfile.bin") {
```
Comment (kelson42) [inline core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt:317]: Why!?! What is ""downloadfile.bin"?!?

File: app/src/androidTest/java/org/kiwix/kiwixmobile/utils/files/FileUtilsTest.kt [hunk 5]
```diff
@@ -90,42 +119,42 @@ class FileUtilsTest {
   }
 
   @Test
-  fun `test decode file name`() {
```
Comment (kelson42) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/utils/files/FileUtilsTest.kt:93]: Don't create one fonction per string to test. Create one test with a array (input string, output string). Please test the specail cases with all the URL encode values.

**Review by kelson42 — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt (hunk 2 above)
Comment (MohitMaliFtechiz) [reply core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt:313]: @kelson42 ,
this documentation is outdated. we are using a new function ```URLUtil.guessFileName()``` of ```android.webkit``` and the updated documentation is available in that class.

File: core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt (hunk 4 above)
Comment (MohitMaliFtechiz) [reply core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt:317]: @kelson42 ,
it is default value from that method, if couldn't get filename from provided url or src.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/utils/files/FileUtilsTest.kt (hunk 5 above)
Comment (MohitMaliFtechiz) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/utils/files/FileUtilsTest.kt]: @kelson42 , done

File: core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt (hunk 2 above)
Comment (kelson42) [reply core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt:313]: If documentation is outdated, you updated it! You don't remove it. Reminder: Such a comment is there to explain:
* The added value (from a user perspective) of this function
* The principle of functionning
* The exception handling
* Anything which is not obvious and necessary to know by using this function

**Review by kelson42 — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt (hunk 4 above)
Comment (kelson42) [reply core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt:317]: This is a fundamental change of the behaviour which not even documented! Actually there is not even a PR description. That's the first problem.

Then I disagree, if there is not Filename then the method should return `undefined` (or worse case empty string), all theses scenario should be part of the automated tests as well.

File: core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt (hunk 2 above)
Comment (MohitMaliFtechiz) [reply core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt:313]: i have added documentation for this function . also updated the PR description.

File: core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt (hunk 4 above)
Comment (MohitMaliFtechiz) [reply core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt]: @kelson42 , I have changed it. if the there is no filename or if ```URLUtils.guessFileName()``` returns the default filename (downloadfile.bin) then this method returns the null value , which is handled in ```downloadFileFromUrl ``` method. i have added unit test for this condition.

**Comment by kelson42** [comment]
@MohitMaliFtechiz you did not requested any new review? You have forgotten?

**Comment by MohitMaliFtechiz** [comment]
> @MohitMaliFtechiz you did not requested any new review? You have forgotten?

@kelson42 Ahh, Sorry my bad, i have forgot to request a re-review from you. previously i have requested a re-review from @gouri-panda .

**Review by kelson42 — APPROVED** [review]

**Pull request merged**